src/bytes.rs

```diff
@@ -1133,6 +1219,43 @@ unsafe fn shared_to_vec(data: &AtomicPtr<()>, ptr: *const u8, len: usize) -> Vec
     shared_to_vec_impl(data.load(Ordering::Relaxed).cast(), ptr, len)
 }
 
+unsafe fn shared_to_mut_impl(shared: *mut Shared, ptr: *const u8, len: usize) -> BytesMut {
+    // The goal is to check if the current handle is the only handle
+    // that currently has access to the buffer. This is done by
+    // checking if the `ref_cnt` is currently 1.
+    //
+    // The `Acquire` ordering synchronizes with the `Release` as
+    // part of the `fetch_sub` in `release_shared`. The `fetch_sub`
+    // operation guarantees that any mutations done in other threads
+    // are ordered before the `ref_cnt` is decremented. As such,
+    // this `Acquire` will guarantee that those mutations are
+    // visible to the current thread.
+    //
+    // Otherwise, we take the other branch, copy the data and call `release_shared``.
+    if (*shared).ref_cnt.load(Ordering::Acquire) == 1 {
+        // Deallocate the `Shared` instance without running its destructor.
+        let shared = *Box::from_raw(shared);
+        let shared = ManuallyDrop::new(shared);
+        let buf = shared.buf;
+        let cap = shared.cap;
+
+        // Rebuild Vec
+        let off = offset_from(ptr, buf);
+        let len = len + off;
+        let v = Vec::from_raw_parts(buf, len, cap);
+
+        BytesMut::from_vec_offset(v, off)
```

Comment:
Does `BytesMut` not also have a vtable with a `Shared` instance? I would prefer to keep the `Shared` instance if possible. Otherwise, the `BytesMut` will need to allocate a new one if it gets promoted.

Reply:
It is not the same `Shared` structure, we would need to "translate" it. I think it can come as a further improvement in a subsequent PR since it is tricky to do. We don't really want the `Bytes` to be aware of `BytesMut` internals (and vice-versa) anyway so I like the `Vec` interface which prevents this coupling.

Reply:
Okay, that's fine. You can keep it as-is.